farm/data_handler/processor.py

```diff
@@ -608,66 +608,98 @@ def file_to_dicts(self, file: str) -> [dict]:
-        tokens = []
-        offsets = []
-        start_of_word = []
-        previous_token_end = -1
-        for token_id, is_special_token, offset in zip(inputs["input_ids"],
-                                                      inputs["special_tokens_mask"],
-                                                      inputs["offset_mapping"]):
-            if not is_special_token:
-                tokens.append(self.tokenizer.convert_ids_to_tokens(token_id))
-                offsets.append(offset[0])
-                start_of_word.append(True if offset[0] != previous_token_end else False)
-                previous_token_end = offset[1]
+        # From here we operate on a per sample basis
+        for dictionary, token_ids, segment_ids in zip(
+                dicts, token_ids_batch, segment_ids_batch
+        ):
 
-        token_dict = {"tokens": tokens,
-                      "offsets": offsets,
-                      "start_of_word": start_of_word}
+            # TODO Build tokenized dict for debug mode
+            tokenized = {}
 
-        if len(token_dict["tokens"]) == 0:
-            logger.warning(f"The following text could not be tokenized, likely because it contains a character that the tokenizer does not recognize: {text}")
-            return []
+            # Truncate results
+            token_ids = self.truncate(token_ids)
+            segment_ids = self.truncate(segment_ids)
 
-        # Build feature dict
-        input_ids, segment_ids = inputs["input_ids"], inputs["token_type_ids"]
+            # Pad results
+            input_ids, padding_mask, segment_ids = self.pad_sequences(token_ids, segment_ids)
+
+            feat_dict = {"input_ids": input_ids,
+                         "padding_mask": padding_mask,
+                         "segment_ids": segment_ids}
+
+            # Create labels
+            if not inference:
+                label_dict = self.generate_labels(dictionary)
+                feat_dict.update(label_dict)
+
+            self.add_basket(dictionary, tokenized, feat_dict)
 
+        if 0 in indices:
+            self._log_samples(2)
+
+        dataset, tensornames = self._create_dataset()
+        ret = [dataset, tensornames]
+        if return_baskets:
+            ret.append(self.baskets)
+        return ret
+
+    def _create_dataset(self):
+        # TODO this is the proposed new version to replace the mother function
+        features_flat = []
+        basket_to_remove = []
+        for basket in self.baskets:
+            if self._check_sample_features(basket):
+                for sample in basket.samples:
+                    features_flat.extend(sample.features)
+            else:
+                # remove the entire basket
+                basket_to_remove.append(basket)
+        dataset, tensor_names = convert_features_to_dataset(features=features_flat)
+        return dataset, tensor_names
+
+    def add_basket(self, dictionary, tokenized, feat_dict):
```

Comment:
I would vote for moving this code directly into `dataset_from_dicts`. It's not that much of code and just reading "add_basket" in the code above is very cryptic.

Reply:
Done in latest commit